File: android/app/src/main/java/cz/drabek/chuvicka/parent/StreamPolicy.kt
package cz.drabek.chuvicka.parent

/** Which stream to ask for. Pure, so the unit tests can run it. The same rule as the iOS app. */
object StreamPolicy {
    data class Inputs(
        val wantsDetail: Boolean,   // The picture is big: landscape, not in the small window.
        val soundOnly: Boolean,     // The sound view, Night mode.
        val thermalHot: Boolean,    // PowerManager thermal status MODERATE or worse.
        val powerSave: Boolean,     // Battery Saver.
        val detailStream: String,
        val everydayStream: String,
    )

    /** The detail (main) stream only for a big picture on a phone that is cool and not saving power. */
    fun detail(i: Inputs): Boolean =
        i.wantsDetail && !i.soundOnly && !i.thermalHot && !i.powerSave && i.detailStream != i.everydayStream

    fun stream(i: Inputs): String = if (detail(i)) i.detailStream else i.everydayStream
}

File: Nursery/Stream/StreamPolicy.swift
import Foundation

/// Which stream to ask for: the detail (main) stream or the everyday (sub) stream.
/// There is no setting for it. Pure, so the tests can run it.
enum StreamPolicy {
    struct Inputs: Equatable {
        /// The screen shows the picture big: zoomed in, full screen, landscape, or an iPad.
        var wantsDetail: Bool
        /// No screen shows the picture: the sound view, Night mode, the background, the lock screen.
        var soundOnly: Bool
        /// The thermal state is serious or critical.
        var thermalHot: Bool
        /// Low Power Mode is on.
        var lowPower: Bool
        var detailStream: String
        var everydayStream: String
    }

    /// True when the detail stream is worth its cost. It has many times the pixels of the
    /// everyday stream, and it keeps the Wi-Fi and the decoder busy.
    static func detail(_ i: Inputs) -> Bool {
        i.wantsDetail && !i.soundOnly && !i.thermalHot && !i.lowPower && i.detailStream != i.everydayStream
    }

    static func stream(_ i: Inputs) -> String { detail(i) ? i.detailStream : i.everydayStream }
}
